rust/bridge/node/futures/src/context.rs

```diff
@@ -0,0 +1,470 @@
+            }
+        });
+
+        if let Some(error) = error_to_throw {
+            cx.throw(error)?;
+        }
+        Ok(())
+    }
+
+    /// Given a function that produces a JavaScript Promise, calls that function and prepares to wait on the result.
+    ///
+    /// This is useful when you want to call an `async` JavaScript function and wait on the result.
+    /// The handling of the result (whether success or failure) is configured using the returned [JsFutureBuilder].
+    ///
+    /// If `promise_callback` fails, that failure will be treated as a rejection to be processed as the promise's result.
+    /// If you want to handle such failures synchronously, you can use [with_context](fn@Self::with_context) instead and manually create a [JsFuture].
+    pub fn await_promise<T>(
+        &self,
+        mut promise_callback: impl for<'a> FnMut(&mut FunctionContext<'a>) -> JsResult<'a, JsObject>,
+    ) -> JsFutureBuilder<T> {
+        let future = self.with_context(|cx| {
+            self.try_catch(cx, |cx| {
+                let promise = promise_callback(cx)?;
+                JsFuture::new(cx, promise, self.clone(), |_cx, _handle| {
+                    panic!("no transform set yet")
+                })
+            })
+        });
+        JsFutureBuilder {
+            async_context: self,
+            state: future,
+        }
+    }
+
+    /// Accesses the single global object that stores all of this context's data.
+    fn context_data_object<'a>(
+        &self,
+        cx: &mut FunctionContext<'a>,
+        create_if_needed: bool,
+    ) -> Handle<'a, JsArray> {
+        let global_key = &self.shared_state.borrow().global_key;
+
+        let context_storage = cx
+            .global()
+            .get(cx, global_key.as_ref())
+            .expect("can always get properties on the global object");
+        if create_if_needed && context_storage.is_a::<JsUndefined>() {
+            let new_storage = cx.empty_array();
+            cx.global()
+                .set(cx, global_key.as_ref(), new_storage)
+                .expect("can always set properties on the global object");
+            new_storage
+        } else {
+            context_storage
+                .downcast()
+                .expect("context data accessed improperly somehow")
+        }
+    }
+
+    /// Stores `value` for the lifetime of this JsAsyncContext.
+    ///
+    /// This allows you to pass data across [FunctionContexts](type@FunctionContext) and thus across `await` boundaries.
+    ///
+    /// The value can be accessed again using [get_context_data](fn@Self::get_context_data).
+    pub fn register_context_data<'a, T: neon::types::Value>(
+        &self,
+        cx: &mut FunctionContext<'a>,
+        value: Handle<'a, T>,
+    ) -> JsAsyncContextKey<T> {
+        let raw_key = self.shared_state.borrow().next_global_id;
+        self.shared_state.update(|state| state.next_global_id += 1);
```

Comment:
Hmm, I didn't think about someone running a *loop* that allocates (and possibly discards) JsAsyncContextKeys. I feel like we're unlikely to have an async entrypoint that stays alive *that* long, but I guess we should at least panic so we don't reuse slots that might still be in use. (We're likely to hit a JavaScript limit long before that, though, since that means we've got a JavaScript object with four billion properties.)